File: core/browser/vision_client.py
# -*- coding: utf-8 -*-
"""Async-клиент для Browser Vision anti-detect API.

Документация: https://docs.browser.vision/ru/about-api

Основные методы:
- list_profiles() — список запущенных профилей
- start_profile(folder_id, profile_id) — запуск профиля → возвращает CDP-порт
- stop_profile(folder_id, profile_id) — остановка профиля
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass

import httpx

logger = logging.getLogger(__name__)


@dataclass(slots=True, frozen=True)
class VisionProfile:
    """Запущенный профиль Vision."""

    folder_id: str
    profile_id: str
    port: int | None


class VisionClient:
    """Async HTTP-клиент для Vision anti-detect браузера."""

    def __init__(
        self,
        x_token: str,
        base_url: str = "http://127.0.0.1:3030",
    ) -> None:
        if not x_token.strip():
            raise RuntimeError("Не задан Vision X-Token")
        self._base = base_url.rstrip("/")
        self._headers = {"X-Token": x_token.strip()}
        self._http = httpx.AsyncClient(timeout=60.0)

    async def list_profiles(self) -> list[VisionProfile]:
        """Получить список запущенных профилей."""
        resp = await self._http.get(
            f"{self._base}/list",
            headers=self._headers,
        )
        resp.raise_for_status()
        data = resp.json()
        logger.info("Vision: ответ /list: %s", data)
        profiles = data.get("profiles") or []
        return [
            VisionProfile(
                folder_id=p["folder_id"],
                profile_id=p["profile_id"],
                port=p.get("port"),
            )
            for p in profiles
        ]

    async def get_profile(self, profile_id: str) -> VisionProfile | None:
        """Возвращает профиль из /list по profile_id."""
        profiles = await self.list_profiles()
        for profile in profiles:
            if profile.profile_id == profile_id:
                return profile
        return None

    async def wait_until_profile_stopped(
        self,
        profile_id: str,
        *,
        timeout_seconds: float = 15.0,
        poll_interval_seconds: float = 1.0,
    ) -> bool:
        """Ждёт, пока профиль исчезнет из списка запущенных."""
        deadline = asyncio.get_running_loop().time() + timeout_seconds
        while True:
            profile = await self.get_profile(profile_id)
            if profile is None:
                return True
            if asyncio.get_running_loop().time() >= deadline:
                return False
            await asyncio.sleep(poll_interval_seconds)

    async def wait_until_profile_has_port(
        self,
        profile_id: str,
        *,
        timeout_seconds: float = 15.0,
        poll_interval_seconds: float = 1.0,
    ) -> int | None:
        """Ждёт, пока у запущенного профиля появится CDP-порт."""
        deadline = asyncio.get_running_loop().time() + timeout_seconds
        while True:
            profile = await self.get_profile(profile_id)
            if profile is not None and profile.port is not None:
                return profile.port
            if asyncio.get_running_loop().time() >= deadline:
                return None
            await asyncio.sleep(poll_interval_seconds)

    async def resolve_folder_id(self, profile_id: str) -> str:
        """Автоматически находит folder_id для profile_id через /list."""
        profiles = await self.list_profiles()
        for p in profiles:
            if p.profile_id == profile_id:
                logger.info(
                    "Vision: folder_id=%s найден для профиля %s",
                    p.folder_id,
                    profile_id,
                )
                return p.folder_id
        raise RuntimeError(
            f"Профиль {profile_id} не найден среди запущенных. "
            f"Запустите его вручную в Vision перед стартом бота."
        )

    async def start_profile(
        self,
        folder_id: str,
        profile_id: str,
        *,
        extra_args: list[str] | None = None,
    ) -> VisionProfile:
        """Запустить профиль и получить CDP-порт для подключения.

        Args:
            folder_id: ID папки в Vision
            profile_id: ID профиля в Vision
            extra_args: дополнительные аргументы браузера

        Returns:
            VisionProfile с заполненным port для CDP-подключения
        """
        url = f"{self._base}/start/{folder_id}/{profile_id}"

        if extra_args:
            # POST с аргументами
            resp = await self._http.post(
                url,
                headers={**self._headers, "Content-Type": "application/json"},
                json={"args": extra_args},
            )
        else:
            # Простой GET-запуск
            resp = await self._http.get(url, headers=self._headers)

        resp.raise_for_status()
        data = resp.json()
        logger.info("Vision: ответ start_profile: %s", data)

        port = data.get("port")

        # Если Vision не вернул порт (профиль уже запущен) — берём из /list
        if port is None:
            logger.info("Vision: порт не в ответе start, пробуем /list")
            port = await self.wait_until_profile_has_port(profile_id)

        profile = VisionProfile(
            folder_id=data.get("folder_id", folder_id),
            profile_id=data.get("profile_id", profile_id),
            port=port,
        )
        logger.info(
            "Vision: профиль %s запущен на порту %s",
            profile.profile_id,
            profile.port,
        )
        return profile

    async def stop_profile(self, folder_id: str, profile_id: str) -> None:
        """Остановить профиль."""
        url = f"{self._base}/stop/{folder_id}/{profile_id}"
        resp = await self._http.get(url, headers=self._headers)
        resp.raise_for_status()
        logger.info("Vision: профиль %s остановлен", profile_id)

    def cdp_url(self, port: int) -> str:
        """Формирует CDP URL для Playwright connectOverCDP."""
        return f"http://127.0.0.1:{port}"

    async def close(self) -> None:
        """Закрытие HTTP-клиента."""
        await self._http.aclose()

File: services/browser-agent/src/vision-client.ts
import type { VisionProfile } from './types.js';

// Таймаут по умолчанию на любой HTTP-вызов к Vision API. Без него зависший (но
// живой) Vision-процесс держал бы сокет открытым → attach/reconnect/maintenance
// висели бы ВЕЧНО, а поллящие циклы (waitUntilProfileHasPort/...) никогда не
// проверили бы собственный deadline (поток стоит на await fetch). См. аудит BA-1.
const DEFAULT_REQUEST_TIMEOUT_MS = 15_000;
// Проба CDP /json/version — короче основного таймаута, т.к. вызывается в поллящем
// цикле: один зависший probe не должен съедать весь deadline ожидания CDP.
const CDP_PROBE_TIMEOUT_MS = 5_000;

/** HTTP-клиент для локального антидетект-браузера Vision на localhost:3030. */
export class VisionClient {
  private readonly baseUrl: string;
  private readonly xToken: string;
  private readonly requestTimeoutMs: number;

  constructor(
    xToken: string,
    baseUrl = 'http://127.0.0.1:3030',
    options?: { requestTimeoutMs?: number },
  ) {
    if (!xToken) throw new Error('Не задан VISION_X_TOKEN');
    this.xToken = xToken;
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.requestTimeoutMs = options?.requestTimeoutMs ?? DEFAULT_REQUEST_TIMEOUT_MS;
  }

  /** fetch с жёстким таймаутом через AbortController. Аборт → понятная ошибка. */
  private async fetchWithTimeout(
    url: string,
    init: RequestInit,
    timeoutMs: number,
    signal?: AbortSignal,
  ): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    const abortFromParent = () => controller.abort(signal?.reason || 'cancelled');
    if (signal?.aborted) {
      abortFromParent();
    } else {
      signal?.addEventListener('abort', abortFromParent, { once: true });
    }
    try {
      return await fetch(url, { ...init, signal: controller.signal });
    } catch (err) {
      if (signal?.aborted) {
        throw new Error('Vision operation cancelled');
      }
      if (controller.signal.aborted) {
        throw new Error(`Vision API ${url} не ответил за ${timeoutMs}ms (timeout)`);
      }
      throw err;
    } finally {
      clearTimeout(timer);
      signal?.removeEventListener('abort', abortFromParent);
    }
  }

  private async request<T>(path: string, signal?: AbortSignal): Promise<T> {
    throwIfAborted(signal);
    const url = `${this.baseUrl}${path}`;
    const res = await this.fetchWithTimeout(
      url,
      { headers: { 'X-Token': this.xToken } },
      this.requestTimeoutMs,
      signal,
    );
    if (!res.ok) {
      throw new Error(`API Vision вернул ошибку ${res.status}: ${await res.text()}`);
    }
    return res.json() as Promise<T>;
  }

  async listProfiles(signal?: AbortSignal): Promise<VisionProfile[]> {
    const data = await this.request<{ profiles: Array<{ folder_id: string; profile_id: string; port: number | null }> }>(
      '/list',
      signal,
    );
    return (data.profiles || []).map((p) => ({
      folder_id: p.folder_id,
      profile_id: p.profile_id,
      port: p.port ?? null,
    }));
  }

  async getProfile(profileId: string, signal?: AbortSignal): Promise<VisionProfile | null> {
    const profiles = await this.listProfiles(signal);
    return profiles.find((p) => p.profile_id === profileId) ?? null;
  }

  async waitUntilProfileStopped(
    profileId: string,
    timeoutSec = 15,
    pollIntervalSec = 1,
    signal?: AbortSignal,
  ): Promise<boolean> {
    const deadline = Date.now() + timeoutSec * 1000;
    while (Date.now() < deadline) {
      throwIfAborted(signal);
      const profile = await this.getProfile(profileId, signal);
      if (!profile) return true;
      await sleep(pollIntervalSec * 1000, signal);
    }
    return false;
  }

  async waitUntilProfileHasPort(
    profileId: string,
    timeoutSec = 15,
    pollIntervalSec = 1,
    signal?: AbortSignal,
  ): Promise<number | null> {
    const deadline = Date.now() + timeoutSec * 1000;
    while (Date.now() < deadline) {
      throwIfAborted(signal);
      const profile = await this.getProfile(profileId, signal);
      if (profile?.port) return profile.port;
      await sleep(pollIntervalSec * 1000, signal);
    }
    return null;
  }

  async waitUntilCdpReady(
    port: number,
    timeoutSec = 20,
    pollIntervalSec = 1,
    signal?: AbortSignal,
  ): Promise<boolean> {
    const deadline = Date.now() + timeoutSec * 1000;
    const versionUrl = `${this.cdpUrl(port)}/json/version`;
    while (Date.now() < deadline) {
      throwIfAborted(signal);
      try {
        const res = await this.fetchWithTimeout(
          versionUrl,
          {},
          CDP_PROBE_TIMEOUT_MS,
          signal,
        );
        if (res.ok) {
          const data = await res.json() as { webSocketDebuggerUrl?: string | null };
          if (typeof data.webSocketDebuggerUrl === 'string' && data.webSocketDebuggerUrl.length > 0) {
            return true;
          }
        }
      } catch {
        throwIfAborted(signal);
        // Пока локальный CDP endpoint не поднялся, просто продолжаем ждать.
      }
      await sleep(pollIntervalSec * 1000, signal);
    }
    return false;
  }

  async resolveFolderId(profileId: string, signal?: AbortSignal): Promise<string> {
    const profile = await this.getProfile(profileId, signal);
    if (!profile) throw new Error(`Профиль ${profileId} не найден в /list`);
    return profile.folder_id;
  }

  async startProfile(
    folderId: string,
    profileId: string,
    options?: {
      portWaitTimeoutSec?: number;
      signal?: AbortSignal;
    },
  ): Promise<VisionProfile> {
    const portWaitTimeoutSec = options?.portWaitTimeoutSec ?? 15;
    const signal = options?.signal;
    const path = `/start/${folderId}/${profileId}`;
    try {
      const data = await this.request<{ port?: number | null; cdp_port?: number | null }>(
        path,
        signal,
      );
      const port = data.port ?? data.cdp_port ?? null;
      if (port) {
        return { folder_id: folderId, profile_id: profileId, port };
      }
      // Порт не вернулся сразу — поллим /list
      const resolvedPort = await this.waitUntilProfileHasPort(
        profileId,
        portWaitTimeoutSec,
        1,
        signal,
      );
      if (resolvedPort) {
        return { folder_id: folderId, profile_id: profileId, port: resolvedPort };
      }
      throw new Error(`Профиль ${profileId} запущен, но CDP-порт не вернулся`);
    } catch (err) {
      throwIfAborted(signal);
      // Запасной вариант: пробуем через GET /list после старта.
      const profile = await this.getProfile(profileId, signal);
      if (profile?.port) return profile;
      throw err;
    }
  }

  async stopProfile(
    folderId: string,
    profileId: string,
    signal?: AbortSignal,
  ): Promise<void> {
    await this.request<void>(`/stop/${folderId}/${profileId}`, signal);
  }

  async restartProfileToRecoverPort(
    folderId: string,
    profileId: string,
    options?: {
      stopTimeoutSec?: number;
      portWaitTimeoutSec?: number;
      settleAfterStopMs?: number;
      signal?: AbortSignal;
    },
  ): Promise<VisionProfile> {
    const stopTimeoutSec = options?.stopTimeoutSec ?? 20;
    const portWaitTimeoutSec = options?.portWaitTimeoutSec ?? 20;
    const settleAfterStopMs = options?.settleAfterStopMs ?? 1_000;
    const signal = options?.signal;

    // Если профиль застрял без CDP-порта, нужен полный stop/start через Vision API.
    await this.stopProfile(folderId, profileId, signal);

    const stopped = await this.waitUntilProfileStopped(
      profileId,
      stopTimeoutSec,
      1,
      signal,
    );
    if (!stopped) {
      throw new Error(
        `Профиль ${profileId} не остановился перед перезапуском для восстановления CDP-порта`,
      );
    }

    // Даем Vision короткую паузу очистить старый процесс перед новым стартом.
    await sleep(settleAfterStopMs, signal);

    return this.startProfile(folderId, profileId, {
      portWaitTimeoutSec,
      signal,
    });
  }

  cdpUrl(port: number): string {
    return `http://127.0.0.1:${port}`;
  }
}

function throwIfAborted(signal?: AbortSignal): void {
  if (signal?.aborted) {
    throw new Error('Vision operation cancelled');
  }
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  throwIfAborted(signal);
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      signal?.removeEventListener('abort', onAbort);
      reject(new Error('Vision operation cancelled'));
    };
    signal?.addEventListener('abort', onAbort, { once: true });
    timer.unref?.();
  });
}
